### PR TITLE
Rewrite presence for performance.

### PR DESCRIPTION
**Pull request by erikjohnston**
- Removes profile information from `m.presence`
- The server idles out users.
- Sending read receipts and typing bumps last active time.
- While a user is online and their `last_active_ago` time remains below 1 minute, don't send updates to client. This is done by including a `currently_active` boolean field that when true indicates that clients should assume the user is only and active until further notice (i.e., their "last active" should be "now"). Clients will receive a new presence event when users stop being active.
- Timeout non-offline states for remote users after 30mins
- Persist the presence state across restarts

TODO:
 - [x] Upload sytest branch
 - [x] Fix unit tests.

### CODE REVIEW

File: synapse/util/wheel_timer.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        Args:
+            now (int): Current time in msec
+            obj (object): Object to be inserted
+            then (int): When to return the object strictly after.
+        """
+        then_key = int(then / self.bucket_size) + 1
+        for entry in self.entries:
+            # Add to first bucket we find. This should gracefully handle inserts
+            # for times in the past.
+            if entry.end_key >= then_key:
```
Comment (NegativeMjark) [inline synapse/util/wheel_timer.py]: If there are no "gaps" in the in the entries list, then why do we need to scan to find the appropriate entry?
Could we simply index into the entry list instead?

File: synapse/storage/util/id_generators.py [hunk 2]
```diff
@@ -130,9 +130,11 @@ def manager():
 
         return manager()
 
-    def get_max_token(self, store):
+    def get_max_token(self, *args):
         """Returns the maximum stream id such that all stream ids less than or
         equal to it have been successfully persisted.
+
+        Used to take a DataStore param, which is no longer needed.
```
Comment (NegativeMjark) [inline synapse/storage/util/id_generators.py]: Could we fix up the call sites rather than adding a *args?

File: synapse/storage/__init__.py [hunk 3]
```diff
@@ -174,6 +197,27 @@ def _get_cache_dict(self, db_conn, table, entity_column, stream_column, max_valu
...
+        """
+
+        sql = (
+            "SELECT user_id, state, last_active_ts, last_federation_update_ts,"
+            " last_user_sync_ts, status_msg, currently_active FROM presence_stream"
+            " WHERE state != ?"
+        )
+        sql = self.database_engine.convert_param_style(sql)
+
+        txn = db_conn.cursor()
```
Comment (NegativeMjark) [inline synapse/storage/__init__.py]: Do you need to close the cursor?

Comment (erikjohnston) [inline synapse/storage/__init__.py]: Done

File: synapse/rest/client/v1/presence.py [hunk 4]
```diff
@@ -35,8 +35,15 @@ def on_GET(self, request, user_id):
...
 
-        state = yield self.handlers.presence_handler.get_state(
-            target_user=user, auth_user=requester.user)
+        if requester.user != user:
+            allowed = yield self.handlers.presence_handler.is_visible(
+                observed_user=user, observer_user=requester.user,
+            )
+
+            if not allowed:
+                raise AuthError(403, "You are allowed to see their presence.")
```
Comment (NegativeMjark) [inline synapse/rest/client/v1/presence.py]: "You are not..."

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**